shotover-proxy/src/server.rs

```diff
@@ -27,9 +29,114 @@ impl<T: Decoder<Item = Messages, Error = anyhow::Error> + Clone + Send> CodecRea
 pub trait CodecWriteHalf: Encoder<Messages, Error = anyhow::Error> + Clone + Send {}
 impl<T: Encoder<Messages, Error = anyhow::Error> + Clone + Send> CodecWriteHalf for T {}
 
+pub trait CodecErrorFixup {
+    /// An method to fix up an error.
+    ///
+    /// Codecs that do not have additional error handling should return `(None, None, Optional<Err>)`
+    ///
+    /// Returns a tuple comprising:
+    ///  * `up_msg` An optional message to send back to the source of the original message.
+    ///  * `down_msg` An optional message to send down chain in the direction the original message was
+    /// flowing.
+    ///  * `an_err` An optional error.  May be the same or a different error.  The Error will be
+    /// processed normally.
+    ///
+    /// # Arguments
+    /// `err` - The original error.
+    ///
+    fn fixup_err(
+        &self,
+        message: Message,
+    ) -> (Option<Message>, Option<Message>, Option<anyhow::Error>);
+}
+
+///
+/// Process the messages.  Any messages with protocol errors are handled by calling the `codec.fixup_err`
+/// method and processing the results.
+/// The resulting Messages contains:
+/// * Any original messages that do not have the protocol_error set.
+/// * Any upstream messages produced by processing the protocol_error message.
+///
+/// Messages that only produce errors for logging or down stream messages are removed from the
+/// result.
+///
+fn handle_protocol_error(
+    codec: &dyn CodecErrorFixup,
+    messages: Messages,
+    tx_out: &UnboundedSender<Messages>,
+) -> Messages {
+    // this code creates a new Vec and uses an iterator with mapping and filtering to
+    // populate it from the original Messages.message Vec.  It may be more efficient to scan the
+    // original Vec and replace or delete individual Message in place.
+    let result = messages
+        .into_iter()
+        .map(|m| {
+            // if there is a protocol error handle it otherwise return the original message.
+            // One thorough the mapping if the message has a protocol_error is dropped.
+            if m.protocol_error != 0 {
+                debug!(
+                    "{:?} processing protocol error: {:?}",
+                    thread::current().id(),
+                    &m
+                );
+                let (up_msg, down_msg, an_err) = codec.fixup_err(m);
+                // if there is a message for upstream send it
+                if let Some(up_msg) = up_msg {
+                    // send up stream messages now
+                    debug!("{:?} Return message: {:?}", thread::current().id(), &up_msg);
+                    tx_out.send(vec![up_msg]).ok();
+                }
```

Comment:
Making the return from `fixup_err` an `enum` would probably make some of this code a bit clearer.